Add Nexus SignalWorkflowOperation

pdoerner opened the pull request:
<!--- Note to EXTERNAL Contributors -->
<!-- Thanks for opening a PR! 
If it is a significant code change, please **make sure there is an open issue** for this. 
We work best with you when we have accepted the idea first before you code. -->

<!--- For ALL Contributors 👇 -->

## What was changed
<!-- Describe what has changed in this PR -->
Added a new type of synchronous Nexus operation for sending signals.

## Why?
<!-- Tell your future self why have you made these changes -->
Convenience and request ID + links handling.

## Checklist
<!--- add/delete as needed --->

1. Closes <!-- add issue number here -->

2. How was this tested:
<!--- Please describe how you tested your changes/how we can test them -->
New functional tests

3. Any docs updates needed?
<!--- update README if applicable
      or point out where to update docs.temporal.io -->

Quinn-With-Two-Ns commented on temporalnexus/operation.go:
```diff
@@ -94,6 +93,20 @@ func NewSyncOperation[I any, O any](
+	signalName string,
+) nexus.Operation[I, nexus.NoValue] {
+	return NewSyncOperation(name, func(ctx context.Context, c client.Client, i I, options nexus.StartOperationOptions) (nexus.NoValue, error) {
+		return nil, c.SignalWorkflow(ctx, workflowID, runID, signalName, i)
```
Don't we need to set the request ID here as well for idempotency

bergundy commented on internal/internal_workflow_client.go:
```diff
@@ -1972,9 +1972,10 @@ func (w *workflowClientInterceptor) SignalWorkflow(ctx context.Context, in *Clie
 		return err
 	}
 
+	links, _ := ctx.Value(NexusOperationRequestIDKey).([]*commonpb.Link)
```
This should be a link key, not the request ID key.

pdoerner commented on internal/interceptor.go:
```diff
@@ -41,6 +41,8 @@ import (
 // Exposed as: [go.temporal.io/sdk/interceptor.Interceptor]
 //
 // Exposed as: [go.temporal.io/sdk/interceptor.Interceptor]
+//
+// Exposed as: [go.temporal.io/sdk/interceptor.Interceptor]
```
I originally aliased `temporalnexus.SignalWorkflowInput` to this type and added it to the comment but then undid that and aliased it to the `internal` type. The automated tools complained about the checks until I ran with `-fix` which I guess just added another line again.

pdoerner replied:
I just went ahead and manually removed all the duplicates

bergundy commented on temporalnexus/operation.go:
```diff
@@ -94,6 +93,28 @@ func NewSyncOperation[I any, O any](
+// NewSignalWorkflowOperation is a helper for creating a synchronous nexus.Operation to deliver a signal.
+//
+// NOTE: Experimental
+func NewSignalWorkflowOperation(name string) nexus.Operation[SignalWorkflowInput, nexus.NoValue] {
```
```suggestion
func NewWorkflowSignalOperation[T any](name string, func(ctx context.Context, input T, options nexus.StartOpertionOptions) SignalWorkflowInput) nexus.Operation[T, nexus.NoValue] {
```

bergundy replied:
The caller doesn't care that this is a signal, that's the handler's call.

bergundy commented on test/nexus_test.go:
```diff
@@ -371,6 +378,56 @@ func TestNexusSyncOperation(t *testing.T) {
+	var out string
+	require.NoError(t, run.Get(ctx, &out))
+	require.Equal(t, "nexus", out)
+
```
Test the request ID and links are properly attached.

bergundy commented on test/nexus_test.go:
```diff
@@ -371,6 +378,56 @@ func TestNexusSyncOperation(t *testing.T) {
 	})
 }
 
+func TestNexusSignalWorkflowOperation(t *testing.T) {
```
Can probably just test from workflow.

bergundy commented on temporalnexus/operation.go:
```diff
@@ -94,6 +93,28 @@ func NewSyncOperation[I any, O any](
 }
 
+// SignalWorkflowInput is the input to a NewSignalWorkflowOperation.
+type SignalWorkflowInput = internal.ClientSignalWorkflowInput
```
Don't reuse the type from the interceptors, you want a type that contains the workflow ID, optional run ID and signal name.

Quinn-With-Two-Ns commented:
Did we test what happens in a failure case like we signal a workflow that does not exist? I would think we would want that to fail the operation?

bergundy reviewed (commented):
This is almost good to go AFAIC.

Something to consider here is that the behavior is different from the WorkflowRunOperation where the default is to pass through the operation input as workflow input. I'm wondering if users will find this inconsistency confusing.

@Quinn-With-Two-Ns you may want to comment on this.

Quinn-With-Two-Ns commented:
>Something to consider here is that the behavior is different from the WorkflowRunOperation where the default is to pass through the operation input as workflow input. I'm wondering if users will find this inconsistency confusing.

I think the difference here is signal needs to know the workflow ID, where as starting a workflow generates the workflow ID. I am not strongly apposed to adding both for signal, I just worry it would encourage a design where uses are passing their workflow IDs in their signal inputs

Quinn-With-Two-Ns commented:
@bergundy I sync'd with the rest of the SDK team and we all agreed that, at first, we just want to expose just the mapping approach.

pdoerner commented:
Planning to upgrade api, nexus SDK, and add handling for returned links in a followup PR. Other than that, I think I addressed all comments. @Quinn-With-Two-Ns and/or @bergundy LMK if this looks good to merge.

Quinn-With-Two-Ns commented on temporalnexus/operation.go:
```diff
@@ -94,6 +93,44 @@ func NewSyncOperation[I any, O any](
 }
 
+// SignalWorkflowInput encapsulates the values required to send a signal to a workflow.
+type SignalWorkflowInput struct {
```
Nit: Experimental
